src/with-unique.js

```diff
@@ -0,0 +1,56 @@
+import { customElements, dashCase, HTMLElement, keys, sym } from './util';
+import { define } from './define';
+
+const _is = sym();
+const baseFunctionStatics = keys(function () {});
+
+let suffix = 0;
+
+function extendStatics (Ctor, Base) {
+  keys(Base).forEach(name => {
+    if (baseFunctionStatics.indexOf(name) > -1) {
+      return;
+    }
+    const descriptor = Object.getOwnPropertyDescriptor(Base, name);
+    if (descriptor) {
+      Object.defineProperty(Ctor, name, descriptor);
+    }
+  });
+}
+
+function formatName (prefix, suffix) {
+  return (
+    prefix.indexOf('-') === -1 ? `x-${prefix}` : prefix
+  ) + (
+    suffix ? `-${suffix}` : ''
+  );
+}
+
+function generateName (Ctor, hint) {
+  const prefix = hint || dashCase(Ctor.name) || 'element';
+  while (customElements.get(formatName(prefix, suffix))) {
+    suffix++;
+  }
+  return formatName(prefix, suffix++);
+}
+
+export function withUnique (Base = HTMLElement) {
+  function Ctor () {
+    const { constructor } = this;
+    define(constructor);
+    return Base.call(this);
```

Comment:
I think if using native ES2015, we'll need to somehow detect that and use `Reflect.construct()` at that point. However, this works now, and we can revisit that later, I think.